crud/common/sharding_key.lua

```diff
@@ -115,6 +116,87 @@ function sharding_key_module.fetch_on_router(space_name)
+function sharding_key_module.is_part_of_pk(space_name, space_format,
+                                           primary_index, sharding_key_def)
+    dev_checks('string', 'table', 'table', 'table')
+
+    if cache.as_index_object[space_name] == nil then
+        cache.as_index_object[space_name] =
+            sharding_key_module.as_index_object(space_format, space_name,
+                                                sharding_key_def)
+    end
+    local sharding_key_as_index_obj = cache.as_index_object[space_name]
+
+    if cache.is_part_of_pk[space_name] ~= nil then
+        return cache.is_part_of_pk[space_name]
+    end
+
+    local is_part_of_pk = true
+    local pk_fieldno_map = utils.get_index_fieldno_map(primary_index.parts)
+    for _, part in ipairs(sharding_key_as_index_obj.parts) do
+        if pk_fieldno_map[part.fieldno] == nil then
+            is_part_of_pk = false
+            break
+        end
+    end
+    cache.is_part_of_pk[space_name] = is_part_of_pk
+
+    return is_part_of_pk
+end
+
+-- Build an array with sharding key values.
+function sharding_key_module.extract_from_index(sharding_key_def,
+                                                index_parts,
+                                                sharding_key_fieldno_map)
+    dev_checks('table', 'table', 'table')
+
+    local sharding_key = {}
+    for i, k in ipairs(sharding_key_def) do
+        local fieldno = index_parts[i].fieldno
+        if sharding_key_fieldno_map[fieldno] == true then
+            table.insert(sharding_key, k)
+        end
+    end
+
+    return sharding_key
+end
+
+-- Extract sharding key from pk.
+-- Returns a table with sharding key or pair of nil and error.
+function sharding_key_module.extract_from_pk(space_obj, sharding_key)
+    dev_checks('table', '?')
+
+    local space_name = space_obj.name
+    local sharding_key_def, err = sharding_key_module.fetch_on_router(space_name)
+    if err ~= nil then
+        return nil, err
+    end
+    if sharding_key_def == nil then
+        return sharding_key
+    end
+
+    local primary_index = space_obj.index[0]
+    local space_format = space_obj:format()
+
+    local is_part_of_pk = sharding_key_module.is_part_of_pk(space_name, space_format,
+                                                            primary_index, sharding_key_def)
+    if is_part_of_pk == false then
+        return nil, ShardingKeyError:new(
+            "Sharding key for space %q is missed in primary index, specify bucket_id",
+            space_name
+        )
+    end
+    if type(sharding_key) ~= 'table' then
+        sharding_key = {sharding_key}
+    end
+    local sharding_key_as_index_obj = cache.as_index_object[space_name]
+    local sharding_key_fieldno_map = utils.get_index_fieldno_map(sharding_key_as_index_obj.parts)
+
+    return sharding_key_module.extract_from_index(sharding_key, primary_index.parts,
+                                                  sharding_key_fieldno_map)
```

Comment:
So we convert it to the index-like object in `is_part_of_pk()` and then convert back there? The first action is cached, I see. But it looks as the code complication due to attempt to use two key definition representation at once. Maybe not so big problem.

Reply:
Moved conversion sharding key definition to index obj to `extract_from_pk()`. I do conversion only once and then use cache only.